radiation/TwoStreamMLPEMod.F90

```diff
@@ -0,0 +1,1716 @@
+    call shr_sys_abort()
+
+  end subroutine endrun
+
+  
+  ! ===============================================================================================
+  
+  subroutine AllocInitTwoStream(this,band_indices,ncan,ncol)
+
+    class(twostream_type) :: this
+    integer               :: band_indices(:)
+    integer               :: ncan
+    integer               :: ncol
+
+    integer :: nbands
+    integer :: ib
+
+    nbands = ubound(band_indices,1)
+
+    allocate(this%n_col(ncan))
+    allocate(this%scelg(ncan,ncol))
+    allocate(this%band(nbands))
+
+    this%n_col(1:ncan)    = unset_int
+    this%n_bands          = nbands
+    this%n_lyr            = ncan
+    this%frac_snow        = unset_r8
+    this%frac_snow_old    = unset_r8
+
+    do ib = 1,nbands
+
+       allocate(this%band(ib)%scelb(ncan,ncol))
+       this%band(ib)%albedo_grnd_diff = unset_r8
+       this%band(ib)%albedo_grnd_beam = unset_r8
+       this%band(ib)%ib = band_indices(ib)
+
+    end do
+
+
+
+    return
+  end subroutine AllocInitTwoStream
+
+  ! ===============================================================================================
+
+  subroutine DeallocTwoStream(this)
+
+    class(twostream_type) :: this
+
+    integer               :: nbands
+    integer               :: ib
+
+    nbands = ubound(this%band,1)
+
+    deallocate(this%scelg)
+    deallocate(this%n_col)
+    do ib = 1,nbands
+       deallocate(this%band(ib)%scelb)
+    end do
+    deallocate(this%band)
+
+    return
+  end subroutine DeallocTwoStream
+
+  ! ===============================================================================================
+
+  subroutine AllocateRadParams(n_pft,n_bands)
+
+    integer,intent(in) :: n_pft
+    integer,intent(in) :: n_bands
+
+    ! Include the zeroth pft index for air
+    
+    allocate(rad_params%rhol(n_bands,n_pft))
+    allocate(rad_params%rhos(n_bands,n_pft))
+    allocate(rad_params%taul(n_bands,n_pft))
+    allocate(rad_params%taus(n_bands,n_pft))
+    allocate(rad_params%xl(n_pft))
+    allocate(rad_params%clumping_index(n_pft))
+
+    allocate(rad_params%phi1(n_pft))
+    allocate(rad_params%phi2(n_pft))
+    allocate(rad_params%kd_leaf(n_pft))
+    allocate(rad_params%kd_stem(n_pft))
+    allocate(rad_params%om_leaf(n_bands,n_pft))
+    allocate(rad_params%om_stem(n_bands,n_pft))
+        
+  end subroutine AllocateRadParams
+
+  ! ================================================================================================
+
+  function GetRdDn(this,ican,icol,ib,vai) result(r_diff_dn)
+
+    class(twostream_type) :: this
+    real(r8),intent(in)   :: vai
+    integer,intent(in)    :: ican
+    integer,intent(in)    :: icol
+    integer,intent(in)    :: ib
+    real(r8)              :: r_diff_dn
+
+    ! Rdn = Ad e−(Kbv) + Re + λ1 B1d e^(av) + λ2 B2d e^(−av)
+
+    associate(scelb => this%band(ib)%scelb(ican,icol), &
+         scelg => this%scelg(ican,icol) )
+
+      r_diff_dn = this%band(ib)%Rbeam_atm*( &
```

Comment:
Yeah, we do it slightly different.  Like you say, instead of the total equals 1, the two components are solved for a unit irradiance of one. That way we can scale the resulting cofficients simply by the actual irradiance once it becomes known. 

I couldn't remember if ED2 had this requirement where its forced to perform a normalized solution, but I remmber it having the normalized solution nonetheless.  With ELM/CLM, the albedo must be retrieved at the beginning of the time-step sequence, so that the atmospheric model can do its calculations, and thus provide the surface irradiance.  And thus, we use the surface irradiance, later in the time-step, to drive energy balance and photosynthesis.  

I can give this whole thing some attention and explain what is going on with comments

Reply:
So just to confirm, you are calling the two stream twice, the first time assuming that all the above-canopy downward irradiance is direct, and a second time when all the above-canopy downward irradiance is assumed diffuse?

Reply:
Just to confirm, this is needed because the host land models need direct and diffuse albedo, correct? I am asking because using a single scaling factor can be computationally more efficient, though it doesn't allow distinguishing direct and diffuse albedo.

Reply:
we do need to provide diffuse and direct albedo